system/src/ble_listening_mode_handler.cpp

```diff
@@ -185,4 +175,91 @@ int BleListeningModeHandler::exit() {
     return SYSTEM_ERROR_NONE;
 }
 
+int BleListeningModeHandler::applyControlRequestConfigurations() {
+    LOG_DEBUG(TRACE, "Apply BLE control request configurations.");
+
+    // Set PPCP (Peripheral Preferred Connection Parameters)
+    hal_ble_conn_params_t ppcp = {};
+    ppcp.size = sizeof(hal_ble_conn_params_t);
+    ppcp.min_conn_interval = BLE_CTRL_REQ_MIN_CONN_INTERVAL;
+    ppcp.max_conn_interval = BLE_CTRL_REQ_MAX_CONN_INTERVAL;
+    ppcp.conn_sup_timeout = BLE_CTRL_REQ_CONN_SUP_TIMEOUT;
+    ppcp.slave_latency = BLE_CTRL_REQ_SLAVE_LATENCY;
+    CHECK(hal_ble_gap_set_ppcp(&ppcp, nullptr));
+
+    // Advertising parameters
+    hal_ble_adv_params_t advParams = {};
+    advParams.size = sizeof(hal_ble_adv_params_t);
+    advParams.type = BLE_ADV_CONNECTABLE_SCANNABLE_UNDIRECRED_EVT;
+    advParams.filter_policy = BLE_ADV_FP_ANY;
+    advParams.interval = BLE_CTRL_REQ_ADV_INTERVAL;
+    advParams.timeout = BLE_CTRL_REQ_ADV_TIMEOUT;
+    advParams.inc_tx_power = false;
+    CHECK(hal_ble_gap_set_advertising_parameters(&advParams, nullptr));
+    CHECK(hal_ble_gap_set_auto_advertise(BLE_AUTO_ADV_ALWAYS, nullptr));
+
+    return SYSTEM_ERROR_NONE;
+}
+
+int BleListeningModeHandler::applyUserAdvData() {
+    CHECK_TRUE((preAdvertising_ || preConnected_), SYSTEM_ERROR_INVALID_STATE);
+    if (!userAdv_) {
+        LOG_DEBUG(TRACE, "Apply user's advertising data set.");
+        CHECK(hal_ble_gap_set_advertising_data(preAdvData_.get(), preAdvDataLen_, nullptr));
+        CHECK(hal_ble_gap_set_scan_response_data(preSrData_.get(), preSrDataLen_, nullptr));
+    }
+    userAdv_ = !userAdv_;
+    return userAdv_ ? SYSTEM_ERROR_NONE : SYSTEM_ERROR_INVALID_STATE;
+}
+
+int BleListeningModeHandler::applyControlRequestAdvData() {
+    LOG_DEBUG(TRACE, "Apply control request advertising data set.");
+    CHECK(hal_ble_gap_set_advertising_data(ctrlReqAdvData_.get(), ctrlReqAdvDataLen_, nullptr));
+    CHECK(hal_ble_gap_set_scan_response_data(ctrlReqSrData_.get(), ctrlReqSrDataLen_, nullptr));
+    return SYSTEM_ERROR_NONE;
+}
+
+int BleListeningModeHandler::enter() {
+    CHECK(constructControlRequestAdvData());
+    CHECK(cacheUserConfigurations());
+    CHECK(applyControlRequestAdvData());
+    CHECK(applyControlRequestConfigurations());
+
+    CHECK(hal_ble_set_callback_on_adv_events(onBleAdvEvents, this, nullptr));
+
+    if (!preAdvertising_ && !preConnected_) {
+        // Start advertising if it is neither connected nor advertising.
+        CHECK(hal_ble_gap_start_advertising(nullptr));
+    }
+    return SYSTEM_ERROR_NONE;
+}
+
+int BleListeningModeHandler::exit() {
+    CHECK(restoreUserConfigurations());
+    CHECK(hal_ble_cancel_callback_on_adv_events(onBleAdvEvents, this, nullptr));
```

Comment:
`preAdvData_`, `preSrData_`, `ctrlReqAdvData_` and `ctrlReqSrData_` need to be deallocated here.

Reply:
In case that we were using `Vector`, the vector should be cleared on exiting listening mode as well.